### PR TITLE
whisper: topic bloom filters

### PR DESCRIPTION
**Pull request by gballet**
In order to offer some measure of plausible deniability, this PR introduces a Bloom filter that topics are inserted in. If a topic is present in the Bloom filter, then the node may be interested in receiving messages with this topic, but will of course receive other messages that they won't be able to decrypt.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: a824ef756e0ac025227e057657037a852ce9f1a0
  * Commits must be prefixed with the package(s) they modify

 * Commit: 4b7eb91a5d186a619d55301b495730b6c51eed06
  * Commits must be prefixed with the package(s) they modify

 * Commit: 062daf3767e6dc911cdb1e0d0ef8e789e2ca0e36
  * Commits must be prefixed with the package(s) they modify

 * Commit: aa3604b665a24a0d3fe51b1f11241cce14f2d9a1
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: whisper/whisperv5/bloom.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+var (
+	errBloomFilterLength = errors.New("The length of the bloom filter must be a power of two between 2 bytes and 4MB")
+	errLengthMismatch    = errors.New("Length of topic and bloom filter don't match")
+	errInvalidTopic      = errors.New("Topic is not a 4 byte array")
+)
+
+// CalculateBloomFilter ceates a bloom filter of `nbytes` bytes with 3 bits set
+// depending on the Keccak256 hash of the topic.
+func CalculateBloomFilter(topic []byte, nbytes uint) (output []byte, err error) {
```
Comment (fjl) [inline whisper/whisperv5/bloom.go]: No need to make this function public. Please remove return value names.

File: whisper/whisperv5/filter.go [hunk 2]
```diff
@@ -69,16 +88,53 @@ func (fs *Filters) Install(watcher *Filter) (string, error) {
...
 	defer fs.mutex.Unlock()
+
+	// Topic removal: regenerate the entire bloom filter, save the topics of
+	// watcher `id`.
+	for wid, watcher := range fs.watchers {
+		if wid != id {
+			for _, topic := range watcher.Topics {
+				if bloom, err := CalculateBloomFilter(topic, topicBloomSize); err == nil {
+					if bloomAddTopics(newBloom, bloom) != nil {
+						fmt.Println("warning, could not add filter %v to whisper. The original list will be kept.", bloom)
```
Comment (fjl) [inline whisper/whisperv5/filter.go]: Please use `log.Warn`

File: whisper/whisperv5/bloom.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	"errors"
+	"math"
+
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var (
+	errBloomFilterLength = errors.New("The length of the bloom filter must be a power of two between 2 bytes and 4MB")
+	errLengthMismatch    = errors.New("Length of topic and bloom filter don't match")
+	errInvalidTopic      = errors.New("Topic is not a 4 byte array")
```
Comment (fjl) [inline whisper/whisperv5/bloom.go]: Error strings should start with a lowercase character.

File: whisper/whisperv5/bloom.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+// CalculateBloomFilter ceates a bloom filter of `nbytes` bytes with 3 bits set
+// depending on the Keccak256 hash of the topic.
+func CalculateBloomFilter(topic []byte, nbytes uint) (output []byte, err error) {
+	var nbits uint = 0    // Number of bits set in the bloom filter
+	var bitsread uint = 0 // Number of bits read so far in the 256-bits hash
+
+	if nbytes == 0 || nbytes > BloomFilterLengthMax {
+		err = errBloomFilterLength
+		return nil, err
```
Comment (fjl) [inline whisper/whisperv5/bloom.go]: `return nil, errBloomFilterLength` is sufficient.

File: whisper/whisperv5/bloom.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	if len(topic) != 4 {
+		err = errInvalidTopic
+		return nil, err
+	}
+
+	// Check that nbytes is a power of two, as it is assumed by the rest of that
+	// function. `bitpitch` is the amount by which the read will advance.
+	bitpitch := uint(math.Floor(math.Log(float64(nbytes*8))/math.Log(2.0) + .5))
```
Comment (fjl) [inline whisper/whisperv5/bloom.go]: Please use `math.Log2(...)` instead of `math.Log(...)/math.Log(2.0)`.

File: whisper/whisperv5/bloom.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+			if int(bitsread+bitpitch) >= len(hash)*8 {
+				hash = crypto.Keccak256(hash)
+				bitsread = 0
+			}
+		}
+	}
+}
+
+// bloomContainsTopics checks if a bloom filter "contains" all the topics that are passed.
+func bloomContainsTopics(bloomFilter []byte, topics []byte) (found bool) {
```
Comment (fjl) [inline whisper/whisperv5/bloom.go]: `topics` here is a slice of bloom filters. That's a bit confusing because `topic` is the actual topic in other places.

**Pull request closed without merging**